lib/course_clone_manager.rb

```diff
@@ -11,7 +11,26 @@ def initialize(course, user, campaign_slug=nil)
 
   def clone!
     @clone = @course.dup
+    course_clone
+    return @clone
+  # If a course with the new slug already exists — an incomplete clone of the
+  # same course — then return the previously-created clone.
+  rescue ActiveRecord::RecordNotUnique
+    return Course.find_by(slug: @clone.slug)
+  end
 
+  def clone_with_assignment!
```

Comment:
Similarly, instead of a new method that basically duplicates the existing one, I think this would be best implemented by extending the existing methods. In this case, you could add a parameter to `clone!` for when the assignments should be copied, and make it default to the current behavior.